tests/test_core.py

```diff
@@ -1859,76 +1869,64 @@ class TestO:
 
         self.assertIsNone(wr_python_object())
 
-    def test_objcinstance_release_owned(self):
-        # Create an object which we own.
-        obj = NSObject.alloc().init()
-
-        # Check that it is marked for release.
-        self.assertTrue(obj._needs_release)
-
-        # Explicitly release the object.
-        obj.release()
-
-        # Check that we no longer need to release it.
-        self.assertFalse(obj._needs_release)
+    def test_objcinstance_created_retained(self):
+        with autoreleasepool():
+            # Create an object which we don't own.
```

Comment:
What's the purpose of the `autoreleasepool` here? It doesn't ensure the object is released, because it doesn't cover the end of the Python wrapper's lifetime.

I think it would make sense to combine the "retained" tests with the "gc_released" tests. Then we'd have two tests, each one covering the full lifetime of an object:

* test_objcinstance_created, based on `alloc`,
* test_objcinstance_retained, based on `stringWithString`

Reply:
> What's the purpose of the autoreleasepool here? It doesn't ensure the object is released, because it doesn't cover the end of the Python wrapper's lifetime.

The autorelease pool is needed to ensure that not only the ObjCInstance is garbage collected (which triggers an autorelease) but also the wrapped Objective-C object is deallocated, which happens only after the pool is emptied. Without it, we get an
```
AssertionError: 2 != 1 : strong property value was not released`.
```
I'll add a comment about this.

> I think it would make sense to combine the "retained" tests with the "gc_released" tests. Then we'd have two tests, each one covering the full lifetime of an object.

I quite like separate test cases because failures will more directly point the relevant problem. But in this case, if the initial retain does not work, both tests will fail anyways, so the benefit is not that large.

Merged those test cases now.